### PR TITLE
Port some algorithms for MPI

### LINKED ISSUE #21032: Port some algorithms for experimental MPI support.
Port algorithm needed for first real-life examples (in particular `SNSPowderDiffraction`?).

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Adds MPI support for a couple of algorithms, all of which are trivial (after some minor modifications), i.e., do not actually require any MPI communication.

Notable changes:
- `MaskBins` now uses `WorkspacePropertyWithIndex`, which is required for MPI support. For compatibility with legacy scripts the old index input field has been kept and forwards to the new one (logging a deprecation warning).
- `WorkspacePropertyWithIndex` now works with workspace validators.

**To test:**

Fixes #21032.
MPI not supported publicly.
Release notes have been updated with non-breaking property change of `MaskBins`.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/MaskBins.cpp [hunk 1]
```diff
@@ -75,40 +61,28 @@ void MaskBins::exec() {
...
-        oss << "One of the workspace indices specified, " << wi
-            << " is above the number of spectra in the workspace (" << numHist
-            << ").";
-        throw std::invalid_argument(oss.str());
-      }
-    }
+  // Copy indices from legacy property
+  std::vector<int> spectraList = this->getProperty("SpectraList");
+  if (!spectraList.empty()) {
+    setProperty("InputWorkspaceIndexSet", spectraList);
```
Comment (OwenArnold) [inline Framework/Algorithms/src/MaskBins.cpp]: You are not preventing users from supplying information to both properties? I think you should probably do so.

File: Framework/Algorithms/src/MaskBins.cpp [hunk 2]
```diff
@@ -119,27 +93,12 @@ void MaskBins::exec() {
       this->findIndices(X, startBin, endBin);
     }
 
-    const int numHists = static_cast<int>(inputWS->getNumberHistograms());
+    const int numHists = static_cast<int>(indexSet.size());
```
Comment (OwenArnold) [inline Framework/Algorithms/src/MaskBins.cpp]: static_cast not required. One of the constructors for `Progress` takes a `size_t`

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Generally looks good

File: Framework/Algorithms/src/MaskBins.cpp [hunk 3]
```diff
@@ -75,40 +61,28 @@ void MaskBins::exec() {
...
-      if ((wi < 0) || (wi >= numHist)) {
-        std::ostringstream oss;
-        oss << "One of the workspace indices specified, " << wi
-            << " is above the number of spectra in the workspace (" << numHist
-            << ").";
-        throw std::invalid_argument(oss.str());
-      }
-    }
+  // Copy indices from legacy property
+  std::vector<int> spectraList = this->getProperty("SpectraList");
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply Framework/Algorithms/src/MaskBins.cpp]: Maybe something to discuss at the developer workshop. If we plan to modernize more algorithms this is certainly a good idea.

[the remark this answers is not in the document]
Comment (OwenArnold) [reply Framework/Algorithms/src/MaskBins.cpp]: Ok sounds like a plan. I'm not going to attend the workshop, so maybe it's something you/lamar could raise.

File: Framework/Algorithms/src/MaskBins.cpp (hunk 1 above)
Comment (OwenArnold) [reply Framework/Algorithms/src/MaskBins.cpp]: You might also consider using [EnableWhenProperty](https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/EnabledWhenProperty.cpp) for completeness in GUI for this algorithm.

**Review by OwenArnold — APPROVED** [review]
Only one suggestion around `EnableWhenProperty` otherwise looks good to go

File: Framework/Algorithms/src/MaskBins.cpp [hunk 4]
```diff
@@ -75,40 +61,31 @@ void MaskBins::exec() {
...
+  if (!spectraList.empty()) {
+    if (!isDefault("InputWorkspaceIndexSet"))
+      throw std::runtime_error("Cannot provide both InputWorkspaceIndexSet and "
```
Comment (OwenArnold) [inline Framework/Algorithms/src/MaskBins.cpp]: See [EnabledWhenProperty](https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/src/EnabledWhenProperty.cpp) for nicer behaviour via the GUI for the algorithm.

**Review by LamarMoore — APPROVED** [review]
This looks good. I like the changes to the declaration of workspace input properties 👍

File: Framework/Algorithms/src/MaskBins.cpp (hunk 4 above)
Comment (LamarMoore) [reply Framework/Algorithms/src/MaskBins.cpp]: @OwenArnold unless you really insist on this I am happy to go ahead and merge

Comment (OwenArnold) [reply Framework/Algorithms/src/MaskBins.cpp]: No I've already approved, so go ahead.

**Pull request merged**